File: backend/app/api/pet.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from core.auth import require_not_observer
from core.exceptions import NotFoundError, BadRequestError
from db.session import get_db
from db import pet as pet_db

router = APIRouter(prefix="/pet", tags=["pet"])

ERROR_MESSAGES = {
    "no_pet": "Питомец ещё не заведён",
    "unknown_item": "Неизвестный предмет",
    "not_owned": "Этого предмета нет в инвентаре",
}


class RenameBody(BaseModel):
    name: str


class EnvironmentBody(BaseModel):
    item_id: str | None = None


@router.get("/")
def get_pet(user_id: int = Depends(require_not_observer), db: Session = Depends(get_db)):
    view = pet_db.get_pet_view(db, user_id)
    if not view and pet_db.has_adopted_but_no_pet(db, user_id):
        pet_db.create_pet_if_needed(db, user_id)
        view = pet_db.get_pet_view(db, user_id)
    if not view:
        raise NotFoundError("Питомец")
    return view


@router.post("/feed/{item_id}")
def feed(item_id: str, user_id: int = Depends(require_not_observer), db: Session = Depends(get_db)):
    ok, error = pet_db.feed_pet(db, user_id, item_id)
    if not ok:
        raise BadRequestError(ERROR_MESSAGES.get(error, error))
    return pet_db.get_pet_view(db, user_id)


@router.post("/play/{item_id}")
def play(item_id: str, user_id: int = Depends(require_not_observer), db: Session = Depends(get_db)):
    ok, error = pet_db.play_with_pet(db, user_id, item_id)
    if not ok:
        raise BadRequestError(ERROR_MESSAGES.get(error, error))
    return pet_db.get_pet_view(db, user_id)


@router.patch("/environment")
def environment(body: EnvironmentBody, user_id: int = Depends(require_not_observer), db: Session = Depends(get_db)):
    ok, error = pet_db.set_environment(db, user_id, body.item_id)
    if not ok:
        raise BadRequestError(ERROR_MESSAGES.get(error, error))
    return pet_db.get_pet_view(db, user_id)


@router.patch("/name")
def rename(body: RenameBody, user_id: int = Depends(require_not_observer), db: Session = Depends(get_db)):
    if not body.name.strip():
        raise BadRequestError("Имя не может быть пустым")
    pet = pet_db.rename_pet(db, user_id, body.name)
    if not pet:
        raise NotFoundError("Питомец")
    return pet_db.get_pet_view(db, user_id)

File: frontend/src/api/pet.ts
import { apiFetch } from './auth'
import type { PetState } from '../types'

const API = import.meta.env.VITE_API_URL

export async function fetchPet(): Promise<PetState | null> {
  const res = await apiFetch(`${API}/pet/`)
  if (res.status === 404) return null
  if (!res.ok) throw new Error('Не удалось загрузить питомца')
  return res.json() as Promise<PetState>
}

export async function feedPet(itemId: string): Promise<PetState> {
  const res = await apiFetch(`${API}/pet/feed/${itemId}`, { method: 'POST' })
  if (!res.ok) {
    const err = await res.json() as { detail?: string }
    throw new Error(err.detail || 'Не удалось покормить')
  }
  return res.json() as Promise<PetState>
}

export async function playWithPet(itemId: string): Promise<PetState> {
  const res = await apiFetch(`${API}/pet/play/${itemId}`, { method: 'POST' })
  if (!res.ok) {
    const err = await res.json() as { detail?: string }
    throw new Error(err.detail || 'Не удалось поиграть')
  }
  return res.json() as Promise<PetState>
}

export async function setPetEnvironment(itemId: string | null): Promise<PetState> {
  const res = await apiFetch(`${API}/pet/environment`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ item_id: itemId }),
  })
  if (!res.ok) {
    const err = await res.json() as { detail?: string }
    throw new Error(err.detail || 'Не удалось сменить фон')
  }
  return res.json() as Promise<PetState>
}

export async function renamePet(name: string): Promise<PetState> {
  const res = await apiFetch(`${API}/pet/name`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    const err = await res.json() as { detail?: string }
    throw new Error(err.detail || 'Не удалось переименовать')
  }
  return res.json() as Promise<PetState>
}
